File: frontend/src/api/backtest.ts
import { apiDelete, apiGet, apiPost, apiPut } from "./client";

export interface HoldingItem {
  ticker: string;
  market: string;
  weight: number;
}

export interface BacktestPortfolioConfig {
  id: string;
  name: string;
  holdings: HoldingItem[];
  created_at: string;
  updated_at: string;
}

export interface BacktestRunRequest {
  portfolio_ids: string[];
  start_date: string;   // "YYYY-MM-DD"
  end_date: string;
  include_spy: boolean;
  include_real_portfolio: boolean;
  reinvest_dividends: boolean;
}

export interface SeriesData {
  name: string;
  values: (number | null)[];  // null = 해당 날짜 데이터 없음 (실제 포트폴리오 시작 전 구간)
}

export interface PortfolioMetrics {
  name: string;
  total_return_pct: number;
  cagr_pct: number;
  mdd_pct: number;
  sharpe_ratio: number;
  volatility_pct: number;
  sortino_ratio: number;
}

export interface BacktestResult {
  dates: string[];
  series: SeriesData[];
  metrics: PortfolioMetrics[];
}

export const fetchBacktestPortfolios = () =>
  apiGet<BacktestPortfolioConfig[]>("/backtest/portfolios");

export const createBacktestPortfolio = (body: { name: string; holdings: HoldingItem[] }) =>
  apiPost<BacktestPortfolioConfig>("/backtest/portfolios", body);

export const updateBacktestPortfolio = (
  id: string,
  body: { name?: string; holdings?: HoldingItem[] },
) => apiPut<BacktestPortfolioConfig>(`/backtest/portfolios/${id}`, body);

export const deleteBacktestPortfolio = (id: string) =>
  apiDelete(`/backtest/portfolios/${id}`);

export const runBacktest = (req: BacktestRunRequest) =>
  apiPost<BacktestResult>("/backtest/run", req);

export interface CorrelationRequest {
  portfolio_ids: string[];
  start_date: string;
  end_date: string;
}

export interface CorrelationResult {
  labels: string[];
  matrix: (number | null)[][];
}

export const runCorrelation = (req: CorrelationRequest) =>
  apiPost<CorrelationResult>("/backtest/correlation", req);

File: backend/app/schemas/backtest.py
"""백테스팅 Pydantic 스키마."""
import uuid
from datetime import date, datetime

from pydantic import BaseModel, field_validator

from app.schemas._validators import validate_portfolio_weights, validate_portfolio_weights_optional


class HoldingItem(BaseModel):
    ticker: str
    market: str
    weight: float  # 0~100, 합계 = 100


class BacktestPortfolioCreate(BaseModel):
    name: str
    holdings: list[HoldingItem]

    @field_validator("holdings")
    @classmethod
    def validate_holdings(cls, v: list[HoldingItem]) -> list[HoldingItem]:
        return validate_portfolio_weights(v)


class BacktestPortfolioUpdate(BaseModel):
    name: str | None = None
    holdings: list[HoldingItem] | None = None

    @field_validator("holdings")
    @classmethod
    def validate_holdings(cls, v: list[HoldingItem] | None) -> list[HoldingItem] | None:
        return validate_portfolio_weights_optional(v)


class BacktestPortfolioResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    name: str
    holdings: list[HoldingItem]
    created_at: datetime
    updated_at: datetime


# ── 백테스팅 실행 ─────────────────────────────────────────

class BacktestRunRequest(BaseModel):
    portfolio_ids: list[uuid.UUID]
    start_date: date
    end_date: date
    include_spy: bool = True
    include_real_portfolio: bool = True
    reinvest_dividends: bool = True

    @field_validator("end_date")
    @classmethod
    def validate_dates(cls, v: date, info) -> date:
        start = info.data.get("start_date")
        if start and v <= start:
            raise ValueError("end_date는 start_date 이후여야 합니다.")
        return v


class SeriesData(BaseModel):
    name: str
    values: list[float | None]  # 기준 100으로 정규화. None = 해당 날짜 데이터 없음 (앞부분 패딩)


class PortfolioMetrics(BaseModel):
    name: str
    total_return_pct: float
    cagr_pct: float
    mdd_pct: float
    sharpe_ratio: float
    volatility_pct: float = 0.0
    sortino_ratio: float = 0.0


class BacktestResult(BaseModel):
    dates: list[str]       # "YYYY-MM-DD" 목록
    series: list[SeriesData]
    metrics: list[PortfolioMetrics]


class CorrelationRequest(BaseModel):
    portfolio_ids: list[uuid.UUID]
    start_date: date
    end_date: date

    @field_validator("end_date")
    @classmethod
    def validate_dates(cls, v: date, info) -> date:
        start = info.data.get("start_date")
        if start and v <= start:
            raise ValueError("end_date는 start_date 이후여야 합니다.")
        return v


class CorrelationResult(BaseModel):
    labels: list[str]
    matrix: list[list[float | None]]
